PapiInstance.cpp

```diff
@@ -1,137 +1,331 @@
+		{ "L1_STM", PAPI_L1_STM },
+		{ "L2_LDM", PAPI_L2_LDM },
+		{ "L2_STM", PAPI_L2_STM },
+		{ "BTAC_M", PAPI_BTAC_M },
+		{ "PRF_DM", PAPI_PRF_DM },
+		{ "L3_DCH", PAPI_L3_DCH },
+		{ "TLB_SD", PAPI_TLB_SD },
+		{ "CSR_FAL", PAPI_CSR_FAL },
+		{ "CSR_SUC", PAPI_CSR_SUC },
+		{ "CSR_TOT", PAPI_CSR_TOT },
+		{ "MEM_SCY", PAPI_MEM_SCY },
+		{ "MEM_RCY", PAPI_MEM_RCY },
+		{ "MEM_WCY", PAPI_MEM_WCY },
+		{ "STL_ICY", PAPI_STL_ICY },
+		{ "FUL_ICY", PAPI_FUL_ICY },
+		{ "STL_CCY", PAPI_STL_CCY },
+		{ "FUL_CCY", PAPI_FUL_CCY },
+		{ "HW_INT", PAPI_HW_INT },
+		{ "BR_UCN", PAPI_BR_UCN },
+		{ "BR_CN", PAPI_BR_CN },
+		{ "BR_TKN", PAPI_BR_TKN },
+		{ "BR_NTK", PAPI_BR_NTK },
+		{ "BR_MSP", PAPI_BR_MSP },
+		{ "BR_PRC", PAPI_BR_PRC },
+		{ "FMA_INS", PAPI_FMA_INS },
+		{ "TOT_IIS", PAPI_TOT_IIS },
+		{ "TOT_INS", PAPI_TOT_INS },
+		{ "INT_INS", PAPI_INT_INS },
+		{ "FP_INS", PAPI_FP_INS },
+		{ "LD_INS", PAPI_LD_INS },
+		{ "SR_INS", PAPI_SR_INS },
+		{ "BR_INS", PAPI_BR_INS },
+		{ "VEC_INS", PAPI_VEC_INS },
+		{ "FLOPS", PAPI_FLOPS },
+		{ "RES_STL", PAPI_RES_STL },
+		{ "FP_STAL", PAPI_FP_STAL },
+		{ "TOT_CYC", PAPI_TOT_CYC },
+		{ "IPS", PAPI_IPS },
+		{ "LST_INS", PAPI_LST_INS },
+		{ "SYC_INS", PAPI_SYC_INS },
+		{ "L1_DCH", PAPI_L1_DCH },
+		{ "L2_DCH", PAPI_L2_DCH },
+		{ "L1_DCA", PAPI_L1_DCA },
+		{ "L2_DCA", PAPI_L2_DCA },
+		{ "L3_DCA", PAPI_L3_DCA },
+		{ "L1_DCR", PAPI_L1_DCR },
+		{ "L2_DCR", PAPI_L2_DCR },
+		{ "L3_DCR", PAPI_L3_DCR },
+		{ "L1_DCW", PAPI_L1_DCW },
+		{ "L2_DCW", PAPI_L2_DCW },
+		{ "L3_DCW", PAPI_L3_DCW },
+		{ "L1_ICH", PAPI_L1_ICH },
+		{ "L2_ICH", PAPI_L2_ICH },
+		{ "L3_ICH", PAPI_L3_ICH },
+		{ "L1_ICA", PAPI_L1_ICA },
+		{ "L2_ICA", PAPI_L2_ICA },
+		{ "L3_ICA", PAPI_L3_ICA },
+		{ "L1_ICR", PAPI_L1_ICR },
+		{ "L2_ICR", PAPI_L2_ICR },
+		{ "L3_ICR", PAPI_L3_ICR },
+		{ "L1_ICW", PAPI_L1_ICW },
+		{ "L2_ICW", PAPI_L2_ICW },
+		{ "L3_ICW", PAPI_L3_ICW },
+		{ "L1_TCH", PAPI_L1_TCH },
+		{ "L2_TCH", PAPI_L2_TCH },
+		{ "L3_TCH", PAPI_L3_TCH },
+		{ "L1_TCA", PAPI_L1_TCA },
+		{ "L2_TCA", PAPI_L2_TCA },
+		{ "L3_TCA", PAPI_L3_TCA },
+		{ "L1_TCR", PAPI_L1_TCR },
+		{ "L2_TCR", PAPI_L2_TCR },
+		{ "L3_TCR", PAPI_L3_TCR },
+		{ "L1_TCW", PAPI_L1_TCW },
+		{ "L2_TCW", PAPI_L2_TCW },
+		{ "L3_TCW", PAPI_L3_TCW },
+		{ "FML_INS", PAPI_FML_INS },
+		{ "FAD_INS", PAPI_FAD_INS },
+		{ "FDV_INS", PAPI_FDV_INS },
+		{ "FSQ_INS", PAPI_FSQ_INS },
+		{ "FNV_INS", PAPI_FNV_INS },
+	};
+
+	auto it = event_map.find(event_code);
+
+	if (it != event_map.end()) {
+		return it->second;
+	} else {
+		return -1;
+	}
+}
+
 void PapiW_start() {
-  printf("Starting Papi Measurement\n");
-  PapiInstance *instance = papi.create();
-  instance->addEvent(PAPI_TOT_INS);
-  instance->addEvent(PAPI_L1_ICM);
-  instance->start();
+	const char *env_codes = getenv("LIB_PAPI_EVENTS");
+
+	PapiInstance *instance = papi.create();
+
+	if (env_codes == nullptr) {
+		std::cout << "LIB_PAPI_EVENTS not defined. Measuring nothing." << std::endl;
```

Comment:
question: would it make more sense to bail out with an error if nothing is specified?

suggestions: add call to `abort()` or add a call to `assert(false && "No PAPI event specified to capture")`.